### ISSUE TITLE
'No main function found' when using undefined function

### ISSUE THREAD

**Issue by systemf-frank** [anoma/juvix#899]
When using an undefined function / variable in the main function, an error is shown that the main function is missing:
```ocaml
mod Undefined-Main where

open Prelude
open Michelson

sig main : int -> int
let main x = nonexisting x
```

When typechecking, it presents us with the error and the complete programs as output:
```
$ stack run -- typecheck -b michelson undefined-main.ju
"No main function found in fromList
    [
        ( "Prelude" :|
            [ "Michelson"
            , "Alias"
            , "*"
            ]
        , RawGFunction
...
...
...
        ( "Prelude" :| [ "ty" ]
        , RawGFunction
            ( RawFunction
                { rawFunName = "Prelude" :| [ "ty" ]
                , rawFunUsage = GOmega
                , rawFunType = Star 1 ()
                , rawFunClauses = RawFunClause
                    { rawClauseTel = []
                    , rawClausePats = []
                    , rawClauseBody = Star 0 ()
                    , rawClauseCatchall = False
                    } :| []
                }
            )
        )
    ]"
```
Of particular interest is that the `Undefined-Main` module is nowhere to be found in the output.

**Comment by systemf-frank** [comment]
Also interesting:
```ocaml
mod Undefined-Main where

open Prelude
open Michelson

sig main : int -> int
let main x = nonexisting -- Note the missing argument x to nonexisting
```

Throws a typechecking error as one would expect: 
```
$ stack run -- typecheck -b michelson undefined-main.ju 
"Error: TypecheckerError
    ( UnboundGlobal
        { unboundGlobal = "nonexisting" :| [] }
    ) on Term: Lam
    ( Elim
        ( Free
            ( Global
                ( "nonexisting" :| [] )
            ) ()
        ) ()
    ) ()"
```

**Comment by systemf-frank** [comment]
Writing the original program in point-free style:
```ocaml
mod Undefined-Main where

open Prelude
open Michelson

sig main : int -> int
let main = nonexisting
```
Gives us the same error, but without program as output:
```
$ stack run -- typecheck -b michelson undefined-main.ju 
"No main function found in fromList []"
```

**Comment by mariari** [comment]
> Writing the original program in point-free style:
> 
> ```ocaml
> mod Undefined-Main where
> 
> open Prelude
> open Michelson
> 
> sig main : int -> int
> let main = nonexisting
> ```
> 
> Gives us the same error, but without program as output:
> 
> ```
> $ stack run -- typecheck -b michelson undefined-main.ju 
> "No main function found in fromList []"
> ```

You can see something similar in the `easypipline` as of now

```haskell
λ> coreify "type Bar : ty = | P bool |  Q bool" (defMichelson)
fromList []
λ> printTimeLapse "type Bar : ty = | P bool |  Q bool" (defMichelson)
[ ( "type" ( "Bar" ":type" "ty" ) () ( "P" "bool" ) ( "Q" "bool" ) ) ]
[ ( "type" ( "Bar" ":type" "ty" ) () ( "P" "bool" ) ( "Q" "bool" ) ) ]
Rec
    { recordContents = T
        { public = fromList
            [
                ( "P", SumCon ( Sum { sumTDef = Nothing, sumTName = "Bar" } ) )
            ,
                ( "Q", SumCon ( Sum { sumTDef = Nothing, sumTName = "Bar" } ) )
            ,
                ( "Bar"
                , TypeDeclar
                    { definitionRepr =
                        (
                            ( "Bar" ":type" "ty" ) ()
                            ( "P" "bool" )
                            ( "Q" "bool" ) ) } ) ]
        , private = fromList [] }
    , recordMTy = Nothing
    , recordOpenList = []
    , recordQualifiedMap = map }
Nothing
Nothing
Nothing
λ> 
```

The odd bit is, I've seen a similar example work when @Acentelles runs the normal pipeline functions. I think the error is the same

**Comment by systemf-frank** [comment]
Just a thought I had and wanted to write it down, haven't checked it yet:
- `main x = nonexisting x` is curried into `main = nonexisting`.
- The typechecker can't find `nonexisting`, so it doesn't show up program (hence the empty list).
- By that, the main function is removed as well, as it lacks a definition.

`main x = noexisting` doesn't have the this problem, as no currying occurs.

**Comment by systemf-frank** [comment]
> > Writing the original program in point-free style:
> > ```ocaml
> > mod Undefined-Main where
> > 
> > open Prelude
> > open Michelson
> > 
> > sig main : int -> int
> > let main = nonexisting
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Gives us the same error, but without program as output:
> > ```
> > $ stack run -- typecheck -b michelson undefined-main.ju 
> > "No main function found in fromList []"
> > ```
> 
> You can see something similar in the `easypipline` as of now
> 
> ```haskell
> λ> coreify "type Bar : ty = | P bool |  Q bool" (defMichelson)
> fromList []
> λ> printTimeLapse "type Bar : ty = | P bool |  Q bool" (defMichelson)
> [ ( "type" ( "Bar" ":type" "ty" ) () ( "P" "bool" ) ( "Q" "bool" ) ) ]
> [ ( "type" ( "Bar" ":type" "ty" ) () ( "P" "bool" ) ( "Q" "bool" ) ) ]
> Rec
>     { recordContents = T
>         { public = fromList
>             [
>                 ( "P", SumCon ( Sum { sumTDef = Nothing, sumTName = "Bar" } ) )
>             ,
>                 ( "Q", SumCon ( Sum { sumTDef = Nothing, sumTName = "Bar" } ) )
>             ,
>                 ( "Bar"
>                 , TypeDeclar
>                     { definitionRepr =
>                         (
>                             ( "Bar" ":type" "ty" ) ()
>                             ( "P" "bool" )
>                             ( "Q" "bool" ) ) } ) ]
>         , private = fromList [] }
>     , recordMTy = Nothing
>     , recordOpenList = []
>     , recordQualifiedMap = map }
> Nothing
> Nothing
> Nothing
> λ> 
> ```
> 
> The Nothings are the pipeline refusing to go further due to some error
> 
> The odd bit is, I've seen a similar example work when @Acentelles runs the normal pipeline functions. I think the error is the same

I think this case is different, as there is no expression at all, just a type definition. i.e. there never was a main to begin with.

**Comment by mariari** [comment]
> > > Writing the original program in point-free style:
> > > ```ocaml
> > > mod Undefined-Main where
> > > 
> > > open Prelude
> > > open Michelson
> > > 
> > > sig main : int -> int
> > > let main = nonexisting
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Gives us the same error, but without program as output:
> > > ```
> > > $ stack run -- typecheck -b michelson undefined-main.ju 
> > > "No main function found in fromList []"
> > > ```
> > 
> > 
> > You can see something similar in the `easypipline` as of now
> > ```haskell
> > λ> coreify "type Bar : ty = | P bool |  Q bool" (defMichelson)
> > fromList []
> > λ> printTimeLapse "type Bar : ty = | P bool |  Q bool" (defMichelson)
> > [ ( "type" ( "Bar" ":type" "ty" ) () ( "P" "bool" ) ( "Q" "bool" ) ) ]
> > [ ( "type" ( "Bar" ":type" "ty" ) () ( "P" "bool" ) ( "Q" "bool" ) ) ]
> > Rec
> >     { recordContents = T
> >         { public = fromList
> >             [
> >                 ( "P", SumCon ( Sum { sumTDef = Nothing, sumTName = "Bar" } ) )
> >             ,
> >                 ( "Q", SumCon ( Sum { sumTDef = Nothing, sumTName = "Bar" } ) )
> >             ,
> >                 ( "Bar"
> >                 , TypeDeclar
> >                     { definitionRepr =
> >                         (
> >                             ( "Bar" ":type" "ty" ) ()
> >                             ( "P" "bool" )
> >                             ( "Q" "bool" ) ) } ) ]
> >         , private = fromList [] }
> >     , recordMTy = Nothing
> >     , recordOpenList = []
> >     , recordQualifiedMap = map }
> > Nothing
> > Nothing
> > Nothing
> > λ> 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The Nothings are the pipeline refusing to go further due to some error
> > The odd bit is, I've seen a similar example work when @Acentelles runs the normal pipeline functions. I think the error is the same
> 
> I think this case is different, as there is no expression at all, just a type definition. i.e. there never was a main to begin with.

The reason why I think it's similar is that in the code i pasted above `ty` isn't defined, and it seems to empty out the dependency ordering phase of `toHR`

```haskell
λ> coreify "sig main : int Prelude.-> int let main x = y" (defMichelson)
fromList [("Prelude" :| ["Michelson",  ... ]
```
is what it should look like being spit back out, but since the list is empty there could be no main, even if it were defined in the frontend

**Comment by mariari** [comment]
After #902 

the `nonexisting` example gives back

```
mod Undefined-Main where

open Prelude
open Michelson

sig main : int -> int
let main x = nonexisting x


stack exec juvix -- compile test/examples/Identity.ju test.out -b "michelson"
"Error on toHR: Name "nonexisting" :| [] not in scope ( probably a bug in the elaborator from frontend )"

```

**Comment by systemf-frank** [comment]
The fix made in #902 slightly improved the error reporting, but I think we can do a lot better than that. Ideally this whole problem would raise an `UnboundGlobal` error, as one would expect. I am keeping this issue open for now.
